cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -0,0 +1,1588 @@
+
+        """
+        Validates that a value from a dictionary is a string and it does not have leading, trailing or double spaces.
+        Adds errors to self.errors if any
+
+        :param str value: The dictionary to validate
+        :param str dict_name: Name of dictionary in the adata (e.g. "uns")
+        :param str key: The key in the dictionary
+
+        :rtype bool
+        :return True if passed, False otherwise
+        """
+
+        is_valid = True
+
+        if not isinstance(value, str):
+            self.errors.append(
+                f"'{value}' in '{dict_name}['{key}']' is not valid, it must be a string."
+            )
+
+            is_valid = False
+        else:
+
+            if value != value.rstrip():
+                self.errors.append(
+                    f"'{value}' in '{dict_name}['{key}']' is not valid, it contains trailing spaces."
+                )
+
+                is_valid = False
+
+            if value != value.lstrip():
+                self.errors.append(
+                    f"'{value}' in '{dict_name}['{key}']' is not valid, it contains leading spaces."
+                )
+
+                is_valid = False
+
+            if value.strip() != " ".join(value.split()):
+                self.errors.append(
+                    f"'{value}' in '{dict_name}['{key}']' is not valid, it contains double spaces."
+                )
+
+                is_valid = False
+
+        return is_valid
+
+    def _validate_enum_in_dict(self, value, enum: List[str], dict_name: str, key: str):
+
+        """
+        Validates that a value from a dictionary is part of a list. Adds errors to self.errors if any
+
+        :param str value: The dictionary to validate
+        :param List[str] enum: The allowed values
+        :param str dict_name: Name of dictionary in the adata (e.g. "uns")
+        :param str key: The key in the dictionary
+
+        :rtype  None
+        """
+
+        if value not in enum:
+            self.errors.append(
+                f"'{value}' in '{dict_name}['{key}']' is not valid. "
+                f"Allowed terms: {enum}."
+            )
+
+    def _validate_X_normalization(self, value):
+
+        """
+        Verifies the adata.uns["X_normalization"] folllows the schema. Adds errors to self.errors if any
+
+        :param value: value in adata.uns["X_normalization"]
+
+        :rtype None
+        """
+
+        if self._validate_str_in_dict(value, "uns", "X_normalization"):
+
+            # If value is "none" then there must NOT be a raw layer
+            if value == "none":
+
+                if self._get_raw_x_loc() == "raw.X":
+                    self.errors.append(
+                        "uns['X_normalization'] is 'none' but 'raw.X' is present. Please indicate "
+                        "the normalization used for 'X'."
+                    )
+
+                if self._get_raw_x_loc() == "raw.X" and not self._is_raw():
```

Comment:
Should this check also be made when `X_norm` _is_ specified?  It's currently nested under the case where `X_norm == 'none'`.

Reply:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203